api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java

```diff
@@ -651,20 +651,92 @@ private void mergeVisits(Patient preferred, Patient notPreferred, PersonMergeLog
+				        if (losing.getEncounters() != null) {
+				            for (Encounter e : losing.getEncounters()) {
+				                e.setPatient(winning.getPatient());
+				                winning.addEncounter(e);
+				            	Encounter persistedEn = encounterService.saveEncounter(e);
+								mergedData.addMovedEncounter(persistedEn.getUuid());
+				            }
+				        }
+				        losing.setEncounters(null); //  manually set the encounters from the non-preferred visit before voiding or all the encounters we just moved will also get voided!
+
+				        visitService.voidVisit(losing, "EMR - Merge Patients: merged into visit " + winning.getVisitId());
+	                     visitService.saveVisit(winning);
+
+	        			break;                   
+	                }
+	                else {
+	                	if (log.isDebugEnabled()) {
+	        				log.debug("Merging visit " + losing.getVisitId() + " to " + winning.getVisitId());
+	        			}
+	        			losing.setPatient(winning.getPatient());
+	        			if (losing.getEncounters() != null) {
+				            for (Encounter e : losing.getEncounters()) {
+				                e.setPatient(winning.getPatient());
+				                Encounter persistedEn = encounterService.saveEncounter(e);
+								mergedData.addMovedEncounter(persistedEn.getUuid());
+				            }
+				        }
+				        
+	        			Visit persisted = visitService.saveVisit(losing);
+	        			mergedData.addMovedVisit(persisted.getUuid());
+	                	break;
+	                }
+	            }   
+	        }        
+        }
+	}
+	private boolean visitsOverlap(Visit v1, Visit v2) {
+        Location where1 = v1.getLocation();
+        Location where2 = v2.getLocation();
+        if ((where1 == null && where2 == null) ||
+                isSameOrAncestor(where1, where2) ||
+                isSameOrAncestor(where2, where1)) {
+            // "same" location, so check if date ranges overlap (assuming startDatetime is never null)
+            return (OpenmrsUtil.compareWithNullAsLatest(v1.getStartDatetime(), v2.getStopDatetime()) <= 0)
+                    && (OpenmrsUtil.compareWithNullAsLatest(v2.getStartDatetime(), v1.getStopDatetime()) <= 0);
+        }
+        return false;
+    }
 	
+	/**
+     * @param a
+     * @param b
+     * @return true if a.equals(b) or a is an ancestor of b.
+     */
+    private boolean isSameOrAncestor(Location a, Location b) {
+        if (a == null || b == null) {
+            return a == null && b == null;
+        }
+        return a.equals(b) || isSameOrAncestor(a, b.getParentLocation());
+    }
+
+	//this will merge remaining encounters that were not merged in merge visit(encounters that don't have visit)
 	private void mergeEncounters(Patient preferred, Patient notPreferred, PersonMergeLogData mergedData) {
 		// change all encounters. This will cascade to obs and orders contained in those encounters
-		// TODO: this should be a copy, not a move
```

Comment:
Did you remove the above todo accidentally?

Reply:
It surely slipped through putting it back